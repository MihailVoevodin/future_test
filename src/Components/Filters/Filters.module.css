.filtrationWrapper {
    display: flex;
    justify-content: center;
    align-items: center;
    margin-top: 10px;
}

.filtrationWrapper select {
    margin-left: 10px;
    width: 100px;
    height: 30px;
}

.filtrationWrapper div:first-child {
    margin-right: 10px;
}

@media (max-width: 400px) {
    .filtrationWrapper span {
        display: block;
        padding-bottom: 10px;
    }
}
