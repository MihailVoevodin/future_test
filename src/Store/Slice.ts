import {createSlice, createAsyncThunk, PayloadAction} from '@reduxjs/toolkit';
import {BooksService} from 'api/BooksService';

export const loadBooksList = createAsyncThunk(
    'books/getBooksList',
    async ({inputValue, startIndex, sorting, filter}: {inputValue: string, startIndex: number, sorting: string, filter: string}, {rejectWithValue, dispatch}) => {
        const response = await BooksService.getBooksList(inputValue, startIndex, sorting, filter);

        if (!response.data.items) {
            dispatch(setIsError())
            dispatch(setIsLoading())
            return;
        }

        if (response.status !== 200) {
            return rejectWithValue('Server Error!');
        }

        return response.data;
    }
);

export const loadPaginateBooksList = createAsyncThunk(
    'books/getPaginateBooksList',
    async ({inputValue, startIndex, sorting, filter}: {inputValue: string, startIndex: number, sorting: string, filter: string}, {rejectWithValue}) => {
        const response = await BooksService.getBooksList(inputValue, startIndex, sorting, filter);

        if (response.status !== 200) {
            return rejectWithValue('Server Error!');
        }

        return response.data;
    }
);

export const loadBook = createAsyncThunk(
    'books/getBook',
    async (id: string | undefined, {rejectWithValue}) => {
        const response = await BooksService.getBook(id);

        if (response.status !== 200) {
            return rejectWithValue('Server Error!');
        }

        return response.data;
    }
);

/**
 * Интерфейс redux-ветки книг.
 * @param inputValue Значение инпута поиска.
 */
export interface IState {
    inputValue: string;
    isLoading: boolean;
    isError: boolean;
    errorMessage: string;
    booksList: IBook[];
    numberOfBooks: number;
    startIndex: number;
    book: Partial<IBook>;
    sorting: string;
    filter: string;
    isPaginateLoad: boolean;
}

const initialState: IState = {
    inputValue: '',
    isLoading: false,
    isError: false,
    errorMessage: '',
    booksList: [],
    numberOfBooks: 0,
    startIndex: 0,
    book: {},
    sorting: 'relevance',
    filter: 'all',
    isPaginateLoad: false,
};

/**
 * Срез книг.
 */
const BooksSlice = createSlice({
    name: 'books',
    initialState: initialState,
    reducers: {
        setInputValue(state, action: PayloadAction<string>) {
            state.inputValue = action.payload;
        },
        setStartIndex(state, action: PayloadAction<number>) {
            state.startIndex = action.payload;
        },
        setBooksList(state) {
            state.booksList = [];
        },
        setBook(state, action) {
            state.book = action.payload;
        },
        setFilter(state, action: PayloadAction<string>) {
            state.filter = action.payload;
        },
        setSorting(state, action: PayloadAction<string>) {
            state.sorting = action.payload;
        },
        setIsError(state) {
            state.isError = true;
            state.errorMessage = 'No matching results';
        },
        setIsLoading(state) {
            state.isLoading = false;
        },
    },
    extraReducers: (builder) => {
        builder
            .addCase(loadBooksList.pending, (state) => {
                state.isLoading = true;
            })
            .addCase(loadBooksList.rejected, (state) => {
                state.isError = true;
                state.errorMessage = 'No matching results';
                state.isLoading = false;
            })
            .addCase(loadBooksList.fulfilled, (state, action) => {
                state.booksList = action.payload.items;
                state.numberOfBooks = action.payload.totalItems;
                state.isLoading = false;
                state.isError = false;
                state.errorMessage = '';
            })
            .addCase(loadPaginateBooksList.pending, (state) => {
                state.isPaginateLoad = true;
            })
            .addCase(loadPaginateBooksList.rejected, (state) => {
                state.isError = true;
                state.errorMessage = 'No matching results';
                state.isPaginateLoad = false;
            })
            .addCase(loadPaginateBooksList.fulfilled, (state, action) => {
                state.booksList.push.apply(state.booksList, action.payload.items);
                state.numberOfBooks = action.payload.totalItems;
                state.isPaginateLoad = false;
                state.isError = false;
                state.errorMessage = '';
            })
            .addCase(loadBook.pending, (state) => {
                state.isLoading = true;
            })
            .addCase(loadBook.rejected, (state) => {
                state.isError = true;
                state.errorMessage = 'No matching results';
                state.isLoading = false;
            })
            .addCase(loadBook.fulfilled, (state, action) => {
                state.book = action.payload;
                state.isLoading = false;
                state.isError = false;
                state.errorMessage = '';
            })
    }
});

export const {
    setInputValue,
    setStartIndex,
    setBooksList,
    setBook,
    setSorting,
    setFilter,
    setIsError,
    setIsLoading,
} = BooksSlice.actions;

export default BooksSlice.reducer;
