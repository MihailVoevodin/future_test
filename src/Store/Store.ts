import {configureStore} from '@reduxjs/toolkit';
import BooksSlice, {IState} from 'Store/Slice'

/**
 * @param books - Ветка книг.
 */
export interface IAppState {
    books: IState;
}

export const store = configureStore<IAppState>({
    reducer: {
        books: BooksSlice,
    },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;