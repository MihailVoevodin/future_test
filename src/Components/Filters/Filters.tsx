import {useAppDispatch, useAppSelector} from 'Common/Hooks';
import {T} from 'Common/Text';
import React, {ChangeEvent} from "react"
import styles from 'Components/Filters/Filters.module.css';
import {loadBooksList, setBooksList, setFilter, setSorting} from 'Store/Slice';

/**
 * Компонент фильтрации и сортировки.
 */
export const Filters: React.FC = () => {
    const dispatch = useAppDispatch()
    const {startIndex, inputValue, sorting, filter} = useAppSelector((state) => state.books);

    const handleChangeFilter = (e: ChangeEvent<HTMLSelectElement>) => {
        dispatch(setBooksList())
        dispatch(setFilter(e.target.value))
        dispatch(loadBooksList({inputValue, startIndex, sorting, filter: e.target.value}))
    }

    const handleChangeSorting = (e: ChangeEvent<HTMLSelectElement>) => {
        dispatch(setBooksList())
        dispatch(setSorting(e.target.value))
        dispatch(loadBooksList({inputValue, startIndex, sorting: e.target.value, filter}))
    }

    return (
        <div className={styles.filtrationWrapper}>
            <div>
                <span>Categories</span>
                <select data-testid="select" name='categories' onChange={handleChangeFilter}>
                    {T.categories.map((category: string) => <option key={category} data-testid="select-option" value={category}>{category}</option>)}
                </select>
            </div>
            <div>
                <span>Sorting by</span>
                <select name='sorting' onChange={handleChangeSorting}>
                    {T.sorting.map((sort: string) => <option key={sort} value={sort}>{sort}</option>)}
                </select>
            </div>
        </div>
    )
}
