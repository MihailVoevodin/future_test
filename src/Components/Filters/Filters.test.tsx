import {fireEvent, render, screen} from '@testing-library/react';
import {Filters} from 'Components/Filters/Filters';
import React from 'react';
import {Provider} from 'react-redux';
import {store} from 'Store/Store';

describe('Filters component', () => {

    test('Filters renders', () => {
        render(
            <Provider store={store}>
                <Filters />
            </Provider>
        );

        expect(screen.getByText(/categories/i)).toBeInTheDocument();
        expect(screen.getByText(/sorting by/i)).toBeInTheDocument();
    })

    test('Filters selection', () => {
        const {getByTestId, getAllByTestId} = render(
            <Provider store={store}>
                <Filters />
            </Provider>
        );

        fireEvent.change(getByTestId('select'), { target: { value: 'art' } })
        let options = getAllByTestId('select-option') as HTMLOptionElement[]
        expect(options[0].selected).toBeFalsy();
        expect(options[1].selected).toBeTruthy();
        expect(options[2].selected).toBeFalsy();
    })

    test('Filters snapshot', () => {
        const filters = render(
            <Provider store={store}>
                <Filters />
            </Provider>
        );

        expect(filters).toMatchSnapshot();
    })
});