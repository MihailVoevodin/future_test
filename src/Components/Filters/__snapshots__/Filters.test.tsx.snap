// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters component Filters snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="filtrationWrapper"
      >
        <div>
          <span>
            Categories
          </span>
          <select
            data-testid="select"
            name="categories"
          >
            <option
              data-testid="select-option"
              value="all"
            >
              all
            </option>
            <option
              data-testid="select-option"
              value="art"
            >
              art
            </option>
            <option
              data-testid="select-option"
              value="biography"
            >
              biography
            </option>
            <option
              data-testid="select-option"
              value="computers"
            >
              computers
            </option>
            <option
              data-testid="select-option"
              value="history"
            >
              history
            </option>
            <option
              data-testid="select-option"
              value="medical"
            >
              medical
            </option>
            <option
              data-testid="select-option"
              value="poetry"
            >
              poetry
            </option>
          </select>
        </div>
        <div>
          <span>
            Sorting by
          </span>
          <select
            name="sorting"
          >
            <option
              value="relevance"
            >
              relevance
            </option>
            <option
              value="newest"
            >
              newest
            </option>
          </select>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="filtrationWrapper"
    >
      <div>
        <span>
          Categories
        </span>
        <select
          data-testid="select"
          name="categories"
        >
          <option
            data-testid="select-option"
            value="all"
          >
            all
          </option>
          <option
            data-testid="select-option"
            value="art"
          >
            art
          </option>
          <option
            data-testid="select-option"
            value="biography"
          >
            biography
          </option>
          <option
            data-testid="select-option"
            value="computers"
          >
            computers
          </option>
          <option
            data-testid="select-option"
            value="history"
          >
            history
          </option>
          <option
            data-testid="select-option"
            value="medical"
          >
            medical
          </option>
          <option
            data-testid="select-option"
            value="poetry"
          >
            poetry
          </option>
        </select>
      </div>
      <div>
        <span>
          Sorting by
        </span>
        <select
          name="sorting"
        >
          <option
            value="relevance"
          >
            relevance
          </option>
          <option
            value="newest"
          >
            newest
          </option>
        </select>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
